File: nbody-simulation/bench.f90
program nbody
    implicit none

    integer, parameter :: n = 1500
    integer, parameter :: steps_warmup = 5
    integer, parameter :: steps = 400
    real(8), parameter :: dt = 0.01d0
    real(8), parameter :: softening = 1.0d-9

    real(8), allocatable :: x(:), y(:), z(:)
    real(8), allocatable :: vx(:), vy(:), vz(:)
    real(8), allocatable :: m(:)
    real(8), allocatable :: fx_buf(:), fy_buf(:), fz_buf(:)

    integer(8) :: seed
    real(8) :: checksum, elapsed_ms
    integer(8) :: count_rate, t1, t2
    integer :: i

    ! Allocate arrays
    allocate(x(n), y(n), z(n))
    allocate(vx(n), vy(n), vz(n))
    allocate(m(n))
    allocate(fx_buf(n), fy_buf(n), fz_buf(n))

    ! Initialize with LCG
    seed = 1
    do i = 1, n
        x(i) = lcg_double(seed)
        y(i) = lcg_double(seed)
        z(i) = lcg_double(seed)
        vx(i) = lcg_double(seed) * 0.1d0
        vy(i) = lcg_double(seed) * 0.1d0
        vz(i) = lcg_double(seed) * 0.1d0
        m(i) = abs(lcg_double(seed)) + 0.5d0
    end do

    ! Warm-up
    call run_steps(steps_warmup, x, y, z, vx, vy, vz, m, fx_buf, fy_buf, fz_buf)

    ! Timed run
    call system_clock(t1, count_rate)
    call run_steps(steps, x, y, z, vx, vy, vz, m, fx_buf, fy_buf, fz_buf)
    call system_clock(t2)

    elapsed_ms = dble(t2 - t1) / dble(count_rate) * 1000.0d0

    ! Compute checksum
    checksum = 0.0d0
    do i = 1, n
        checksum = checksum + x(i) + y(i) + z(i) + vx(i) + vy(i) + vz(i)
    end do

    write(*, '(A,F0.3,A,F0.6)') 'elapsed_ms=', elapsed_ms, ' checksum=', checksum

    ! Deallocate
    deallocate(x, y, z, vx, vy, vz, m, fx_buf, fy_buf, fz_buf)

contains

    function lcg_double(state) result(val)
        integer(8), intent(inout) :: state
        real(8) :: val
        integer(8) :: shifted
        integer(8), parameter :: mult = int(Z'5851F42D4C957F2D', 8)

        ! state = state * mult + 1 (with wraparound)
        state = state * mult + 1_8
        
        ! Logical shift right by 11
        shifted = ishft(state, -11)
        ! Ensure high bits are zero (already true for logical shift, but being explicit)
        shifted = iand(shifted, int(Z'1FFFFFFFFFFFFF', 8))

        val = dble(shifted) * (1.0d0 / 9007199254740992.0d0) * 2.0d0 - 1.0d0
    end function lcg_double

    subroutine run_steps(count, x, y, z, vx, vy, vz, m, fx_buf, fy_buf, fz_buf)
        integer, intent(in) :: count
        real(8), intent(inout) :: x(:), y(:), z(:)
        real(8), intent(inout) :: vx(:), vy(:), vz(:)
        real(8), intent(in) :: m(:)
        real(8), intent(inout) :: fx_buf(:), fy_buf(:), fz_buf(:)

        integer :: step, i, j
        real(8) :: xi, yi, zi, fxi, fyi, fzi, mi
        real(8) :: dx, dy, dz, dist2, inv, inv3, s_i, s_j

        do step = 1, count
            ! Zero out force buffers
            fx_buf = 0.0d0
            fy_buf = 0.0d0
            fz_buf = 0.0d0

            do i = 1, n
                xi = x(i)
                yi = y(i)
                zi = z(i)
                fxi = fx_buf(i)
                fyi = fy_buf(i)
                fzi = fz_buf(i)
                mi = m(i)

                do j = i + 1, n
                    dx = x(j) - xi
                    dy = y(j) - yi
                    dz = z(j) - zi
                    dist2 = dx * dx + dy * dy + dz * dz + softening
                    inv = 1.0d0 / sqrt(dist2)
                    inv3 = inv * inv * inv

                    s_i = m(j) * inv3
                    s_j = mi * inv3

                    fxi = fxi + dx * s_i
                    fyi = fyi + dy * s_i
                    fzi = fzi + dz * s_i

                    fx_buf(j) = fx_buf(j) - dx * s_j
                    fy_buf(j) = fy_buf(j) - dy * s_j
                    fz_buf(j) = fz_buf(j) - dz * s_j
                end do

                fx_buf(i) = fxi
                fy_buf(i) = fyi
                fz_buf(i) = fzi
            end do

            ! Update velocities
            do i = 1, n
                vx(i) = vx(i) + dt * fx_buf(i)
                vy(i) = vy(i) + dt * fy_buf(i)
                vz(i) = vz(i) + dt * fz_buf(i)
            end do

            ! Update positions
            do i = 1, n
                x(i) = x(i) + dt * vx(i)
                y(i) = y(i) + dt * vy(i)
                z(i) = z(i) + dt * vz(i)
            end do
        end do
    end subroutine run_steps

end program nbody

File: nbody-simulation/Bench.java
import java.util.Random;

class LCG {
    private long state;

    public LCG(long seed) {
        this.state = seed;
    }

    public long nextU64() {
        state = (state * 6364136223846793005L + 1L);
        return state;
    }

    public double nextDouble() {
        long v = nextU64();
        // Map to [-1.0, 1.0] using the same logic as C/Rust
        // We use (v >>> 11) to get 53 bits of precision, same as a double's mantissa
        return ((v >>> 11) * (1.0 / 9007199254740992.0)) * 2.0 - 1.0;
    }
}

public class Bench {
    static void runSteps(int n, int count, double dt, double softening,
                        double[] x, double[] y, double[] z,
                        double[] vx, double[] vy, double[] vz,
                        double[] m,
                        double[] fxBuf, double[] fyBuf, double[] fzBuf) {
        
        for (int step = 0; step < count; step++) {
            // Zero out force buffers
            for (int i = 0; i < n; i++) {
                fxBuf[i] = 0.0;
                fyBuf[i] = 0.0;
                fzBuf[i] = 0.0;
            }

            for (int i = 0; i < n; i++) {
                double xi = x[i];
                double yi = y[i];
                double zi = z[i];
                double fxi = fxBuf[i];
                double fyi = fyBuf[i];
                double fzi = fzBuf[i];
                double mi = m[i];

                for (int j = i + 1; j < n; j++) {
                    double dx = x[j] - xi;
                    double dy = y[j] - yi;
                    double dz = z[j] - zi;
                    double dist2 = dx * dx + dy * dy + dz * dz + softening;
                    double inv = 1.0 / Math.sqrt(dist2);
                    double inv3 = inv * inv * inv;
                    
                    double s_i = m[j] * inv3;
                    double s_j = mi * inv3;

                    fxi += dx * s_i;
                    fyi += dy * s_i;
                    fzi += dz * s_i;

                    fxBuf[j] -= dx * s_j;
                    fyBuf[j] -= dy * s_j;
                    fzBuf[j] -= dz * s_j;
                }
                fxBuf[i] = fxi;
                fyBuf[i] = fyi;
                fzBuf[i] = fzi;
            }

            for (int i = 0; i < n; i++) {
                vx[i] += dt * fxBuf[i];
                vy[i] += dt * fyBuf[i];
                vz[i] += dt * fzBuf[i];
            }

            for (int i = 0; i < n; i++) {
                x[i] += dt * vx[i];
                y[i] += dt * vy[i];
                z[i] += dt * vz[i];
            }
        }
    }

    public static void main(String[] args) {
        final int n = 1500;
        final int stepsWarmup = 5;
        final int steps = 400;
        final double dt = 0.01;
        final double softening = 1e-9;

        double[] x = new double[n];
        double[] y = new double[n];
        double[] z = new double[n];
        double[] vx = new double[n];
        double[] vy = new double[n];
        double[] vz = new double[n];
        double[] m = new double[n];
        
        double[] fxBuf = new double[n];
        double[] fyBuf = new double[n];
        double[] fzBuf = new double[n];

        LCG lcg = new LCG(1);
        for (int i = 0; i < n; i++) {
            x[i] = lcg.nextDouble();
            y[i] = lcg.nextDouble();
            z[i] = lcg.nextDouble();
            vx[i] = lcg.nextDouble() * 0.1;
            vy[i] = lcg.nextDouble() * 0.1;
            vz[i] = lcg.nextDouble() * 0.1;
            m[i] = Math.abs(lcg.nextDouble()) + 0.5;
        }

        // Warm-up
        runSteps(n, stepsWarmup, dt, softening, x, y, z, vx, vy, vz, m, fxBuf, fyBuf, fzBuf);

        long startMs = System.currentTimeMillis();
        runSteps(n, steps, dt, softening, x, y, z, vx, vy, vz, m, fxBuf, fyBuf, fzBuf);
        long endMs = System.currentTimeMillis();

        double checksum = 0.0;
        for (int i = 0; i < n; i++) {
            checksum += x[i] + y[i] + z[i] + vx[i] + vy[i] + vz[i];
        }

        System.out.printf("elapsed_ms=%.3f checksum=%.6f\n", (double)(endMs - startMs), checksum);
    }
}
